corehq/apps/domain/auth.py

```diff
@@ -19,53 +19,48 @@
 TOKEN = 'token'
 
 
-def determine_authtype_from_header(request, default=DIGEST):
+def determine_authtype_from_header(request, support_non_j2me_digest=True):
     """
     Guess the auth type, based on the headers found in the request.
+
+    CommCare mobile typically sends an unauthenticated request first, and
+    expects a basic auth challenge in response. This is at odds with the spec
+    for digest auth, in which the client sends an unauthenticated response and
+    receives a digest auth challenge. This means we can't support both CommCare
+    mobile and digest auth at the same endpoint.
```

Comment:
I'm not sure I understand how this statement can be true if the mobile endpoints support both. You're saying they don't support "real" digest? Just this bizarre commcare mobile digest frankenstein thing?

Reply:
I think this comment is also misleading. I dont think the phone does anything wrong. Sending an unauthenticated request, receiving for a 401, then attempting to authenticate in the method requested is I believe a very normal request cycle. The issue is determining what the correct response from the server is. When we receive an unauthenticated request do we prompt for basic or digest? The issue isn't really mobile or digest, it is clients that can only auth by one method and do not tell us that method. It seems as wrong to assume an unauthenticated request is really a request for digest auth as it is to assume it is a request for basic auth. The crux of the matter is that in instances when the client doesn't tell us, we can only support one type of auth and we just need to decide what that is and be explicit about it. In order to help make that decision we can use the user agent, like we do for j2me, or a header, like the authorization ones below, but I think it still makes sense to think of those as a default for when we are not sure.

Reply:
Ok, thanks for these clarifications - I think I have a much better sense of why this is complex/hard now (and sorry it took this long for me to get there).

Thinking about the current situation it seems like it's something like this:

| Default Challenge | Supports Digest? | Supports Basic? | Needed For |
| --- | --- | --- | --- | 
| Basic | Via non-standard header / urlparam only | Yes | Mobile Endpoints |
| Digest | Yes | Via standard auth header or nonstandard user agent header/urlparam | ??? |

And I think what you've found is that for obscure legacy support reasons we now find ourselves in a situation where we have different endpoints that need to be configured in both of the above ways due to reasons that a) live outside our control, and b) aren't particularly sensical. Does that sound right? And can you remind me what the requirement is for the bottom line where the "???"s are?

I think maybe we could just create two sensibly-named methods to represent these two configurations and then not change the inner code (and just make it private and well documented essentially)? I do agree with @calellowitz about the default argument to this particular function, despite the odd behavior that it creates.

Finally, (and sorry this probably isn't the right place for this comment) - is there any reason not to support sessions/tokens/etc. in all endpoints? Then I think we can make it so that we _only_ need these two decorators.

Reply:
That's my understanding, yes.  Thanks for the coherent summary!

The second row above is needed for API endpoints - both our standard tastypie stuff and custom/one-off code.  It's also not clear exactly which of those we do need for a given endpoint.  We could certainly require API consumers to use a particular type of auth, although I'm guessing we have a variety of auth types currently in use, so that sounds hard to change.

> Finally, (and sorry this probably isn't the right place for this comment) - is there any reason not to support sessions/tokens/etc. in all endpoints?

No idea.

This all sounds good.  I'll roll back some of my overzealous changes and rename these two decorators to something sensible.  I think I'll also keep the third one with token auth separate for now and consider that a separate piece of tech debt to clean up as @emord described.  For the two main ones, how about:
`@mobile_auth`, `@api_auth`
Other possibility:
`@require_mobile_login`, `@require_api_login`
Also open to suggestions - not really particular on the exact names.

Reply:
> Finally, (and sorry this probably isn't the right place for this comment) - is there any reason not to support sessions/tokens/etc. in all endpoints?

I think @sravfeyn removed session auth on purpose from all the api endpoints for one of the security audits we had to do for ICDS/eNikshay. I am not sure of the rationale but I believe we do not want session auth on all endpoints. I see no reason token auth should not be.

The rest sounds great

Reply:
>  We could certainly require API consumers to use a particular type of auth, although I'm guessing we have a variety of auth types currently in use, so that sounds hard to change.

i could be way off in this but my understanding was that most clients could handle either type of auth. Things like curl, or any programming language, should not only be able to do both, but any solid http library should handle both transparently without bothering the programmer. That in fact is why we had this problem to begin with, the phone was willing to do whatever auth we asked of it, so it switched from basic to digest without anyone knowing, it just had a bug in its digest implementation. Which is all to say, im sure we will break someone's workflow because what doesn't, but I think this may be less of a drastic break than we expect, and whatever eNikshay partner required digest and could not basic auth, may be more the outlier than the rule.

Reply:
@calellowitz good point.  And in the case of enikshay, it's a custom endpoint anyways, so it could also support _only_ digest auth if we wanted.

Reply:
I imagine there's a world in which we go through a lot of trouble to remove those requirements via proactive comms, soft asserts, and support, but wonder whether the ROI is worth it. living in a two-decorator world seems like it might not be too bad